File: nodejs/backend/src/service/UserService.ts
import { UserRepository } from '../repository/UserRepository';
import { User } from '../model/User';
import { UserNotFoundError } from '../error/AppError';

export class UserService {
  constructor(private readonly userRepo: UserRepository) {}

  async list(): Promise<User[]> {
    return this.userRepo.findAll();
  }

  async delete(id: string): Promise<void> {
    const deleted = await this.userRepo.deleteById(id);
    if (!deleted) {
      throw new UserNotFoundError();
    }
  }
}

File: rust/backend/src/service/user_service.rs
use crate::{error::AppError, model::User, repository::UserRepository};
use std::sync::Arc;
use uuid::Uuid;

#[derive(Clone)]
pub struct UserService {
    user_repo: Arc<dyn UserRepository>,
}

impl UserService {
    pub fn new(user_repo: Arc<dyn UserRepository>) -> Self {
        Self { user_repo }
    }

    pub async fn list(&self) -> Result<Vec<User>, AppError> {
        Ok(self.user_repo.find_all().await?)
    }

    pub async fn delete(&self, id: Uuid) -> Result<(), AppError> {
        match self.user_repo.delete(id).await? {
            Some(_) => Ok(()),
            None => Err(AppError::UserNotFound),
        }
    }
}
